toolbox/relnotes/main.go

```diff
@@ -0,0 +1,784 @@
+		// due to the recently expanded blocking test list for master. The expanded test
+		// list is often unable to find a complete passing set and find_green_build runs
+		// unbounded for hours
+		extraFlag = "--limit=30"
+	}
+
+	f, err := os.OpenFile(outputFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err = f.Close(); err != nil {
+			result = fmt.Errorf("failed to close file %s, %s", outputFile, err)
+		}
+	}()
+
+	f.WriteString(fmt.Sprintf("## State of %s branch\n", branch))
+
+	// Call script find_green_build to get CI job status
+	content, err := u.Shell(os.Getenv("GOPATH")+"/src/k8s.io/release/find_green_build", "-v", extraFlag, branch)
+	if err == nil {
+		f.WriteString(fmt.Sprintf("%sGOOD TO GO!%s\n\n", green, off))
+	} else {
+		f.WriteString(fmt.Sprintf("%sNOT READY%s\n\n", red, off))
+	}
+
+	f.WriteString("### Details\n```\n")
+	f.WriteString(content)
+	f.WriteString("```\n")
+
+	log.Printf("CI job status fetched.")
+	return result
+}
+
+// createBody creates the general documentation, example and downloads table body for the
+// markdown file.
+func createBody(f *os.File, releaseTag, branch, docURL, exampleURL, releaseTars string) error {
+	var title string
+	if *preview {
+		title = "Branch "
+	}
+
+	if releaseTag == "HEAD" || releaseTag == branchHead {
+		title += branch
+	} else {
+		title += releaseTag
+	}
+
+	if *preview {
+		f.WriteString(fmt.Sprintf("**Release Note Preview - generated on %s**\n", time.Now().Format("Mon Jan  2 15:04:05 MST 2006")))
+	}
+
+	f.WriteString(fmt.Sprintf("\n# %s\n\n", title))
+	f.WriteString(fmt.Sprintf("[Documentation](%s) & [Examples](%s)\n\n", docURL, exampleURL))
+
+	if releaseTars != "" {
+		f.WriteString(fmt.Sprintf("## Downloads for %s\n\n", title))
+		err := createDownloadsTable(f, releaseTag, "", releaseTars+"/kubernetes.tar.gz", releaseTars+"/kubernetes-src.tar.gz")
+		if err != nil {
+			return fmt.Errorf("failed to create downloads table: %s", err)
+		}
+		err = createDownloadsTable(f, releaseTag, "Client Binaries", releaseTars+"/kubernetes-client*.tar.gz")
+		if err != nil {
+			return fmt.Errorf("failed to create downloads table: %s", err)
+		}
+		err = createDownloadsTable(f, releaseTag, "Server Binaries", releaseTars+"/kubernetes-server*.tar.gz")
+		if err != nil {
+			return fmt.Errorf("failed to create downloads table: %s", err)
+		}
+		err = createDownloadsTable(f, releaseTag, "Node Binaries", releaseTars+"/kubernetes-node*.tar.gz")
+		if err != nil {
+			return fmt.Errorf("failed to create downloads table: %s", err)
+		}
+		f.WriteString("\n")
+	}
+	return nil
+}
+
+// createDownloadTable creates table of download link and sha256 hash for given file.
+func createDownloadsTable(f *os.File, releaseTag, heading string, filename ...string) error {
+	var urlPrefix string
+
+	if *releaseBucket == "kubernetes-release" {
+		urlPrefix = k8sReleaseURLPrefix
+	} else {
```

Comment:
What if *releaseBucket == "" ?

Reply:
The flag has default value "kubernetes-release" and user can specify alternative google storage bucket to point to in generated notes. The download-table-creation logic is Kubernetes-specified and doesn't apply to other projects yet.

Reply:
What if user specifies the flag as ""? We will get `https://storage.googleapis.com//release`

Reply:
Yes, I think it only affect the URL text displayed in generated notes (the program doesn't actually try to fetch resources from that URL), and it's user's responsibility to provide correct address. I don't know if there are cases other than `https://dl.k8s.io` and `https://storage.googleapis.com/<some bucket>/release`. In that case user may manually modify the generated text.

I feel like we should keep the program running even if user specifies the flag as "". We can have the program log an error message when `*releaseBucket == ""`, but it won't cover cases if user specifies some invalid non-existing google storage bucket.

Reply:
agreed :)